### PR TITLE
Prologix settings are configurable.

### PR DESCRIPTION
**Pull request by bmoneke**
Incorporate properties from https://github.com/thosou/pymeasure/blob/dev/prologix-ethernet/pymeasure/adapters/prologix.py by @thosou

given in #172 

Only the `read` methods differ (++read vs ++read eoi), could you shed light on this question @msmttchr ?

### CODE REVIEW

File: pymeasure/adapters/prologix.py [hunk 1]
```diff
@@ -55,6 +55,9 @@ class PrologixAdapter(VISAAdapter):
         .. deprecated:: 0.11
             Implement it in the instrument's `read` method instead.
 
+    :param auto: Default to 0 to turn off read-after-write and address instrument to listen.
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:58]:
```suggestion
    :param auto: Enable (1) or disable (0)  read-after-write and address instrument to listen.
```

File: pymeasure/adapters/prologix.py [hunk 2]
```diff
@@ -55,6 +55,9 @@ class PrologixAdapter(VISAAdapter):
...
+    :param eoi: Enable (1) or disable (0) EOI assertion.
+    :param eos: Set command termination 0 for CR+LF; 2 for LF.
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:60]:
```suggestion
    :param eos: Set command termination 0 for CR+LF; 1 for CR, 2 for LF, 3 for none.
```

File: pymeasure/adapters/prologix.py [hunk 3]
```diff
@@ -107,15 +110,66 @@ def __init__(self, resource_name, address=None, rw_delay=0, serial_timeout=None,
...
+            self.auto = auto
+            self.eoi = eoi
+            self.eos = eos
+
+    @property
+    def auto(self):
+        """Configure Prologix GPIB controller to automatically address instruments
+        to talk after sending them a command in order to read their response. The
+        feature called, Read-After-Write, saves the user from having to issue read commands
+        repeatedly. This property enabled or disabled the Read-After-Write feature.
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:122]:
```suggestion
        repeatedly. This property enables or disables this feature.
```
I think also valid inputs values and their meaning should be specified in the doc string.

File: pymeasure/adapters/prologix.py [hunk 4]
```diff
@@ -107,15 +110,66 @@ def __init__(self, resource_name, address=None, rw_delay=0, serial_timeout=None,
...
+        """
+        self.write("++auto")
+        return int(self.read())
+
+    @auto.setter
+    def auto(self, value):
+        self.write("++auto {}".format(value))
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:129]:
```suggestion
        self.write(f"++auto {value}")
```
I think we mostly used this compact form in the code, hence the suggestion. If agreed there are also few other cases below.

File: pymeasure/adapters/prologix.py [hunk 5]
```diff
@@ -107,15 +110,66 @@ def __init__(self, resource_name, address=None, rw_delay=0, serial_timeout=None,
...
+
+    @property
+    def eoi(self):
+        """Enables or Disables the assertion of the EOI signal with the last character
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:133]: I think also valid inputs values and their meaning should be specified in the doc string.

File: pymeasure/adapters/prologix.py [hunk 6]
```diff
@@ -107,15 +110,66 @@ def __init__(self, resource_name, address=None, rw_delay=0, serial_timeout=None,
...
+    def eoi(self, value):
+        self.write("++eoi {}".format(value))
+
+    @property
+    def eos(self):
+        """Specify GPIB termination characters. When data from host is received,
+        all non-escaped LF, CR and ESC characters are removed and GPIB
+        terminators, as specified by this command, are appended before sending the data to
+        instruments. This command does not affect data from instruments received over GPIB
+        port.
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:150]: I think also valid inputs values and their meaning should be specified in the doc string.

**Review by msmttchr — COMMENTED** [review]
Some minor comments on the docstrings.

File: pymeasure/adapters/prologix.py [hunk 7]
```diff
@@ -191,6 +245,7 @@ def _read(self, **kwargs):
         :param kwargs: Keyword arguments for the connection itself.
         :returns str: ASCII response of the instrument (excluding read_termination).
         """
+        # TODO Prologix Ethernet version has only `write("++read")`
```
Comment (msmttchr) [inline pymeasure/adapters/prologix.py:248]: You mean in the code of @thosou ?

**Comment by msmttchr** [comment]
> Only the `read` methods differ (++read vs ++read eoi), could you shed light on this question @msmttchr ?

Not sure, I would think is linked to instrument behaviour and so I have no valid explanation.
It is possible that this change was unintentional and left over after debugging.

**Comment by bmoneke** [comment]
Thanks for your review @msmttchr 
I changed the docstrings to pymeasure style ("Control", "Get"...), at the same time, I made the properties accept useful settings (True/False or a termination string).

As I wrote the tests, I noticed, that I have to adjust the `read` method as well: I think, if you try to read the auto setting with `++auto`, you do not have to send `++read eoi`, don't you? Not the read method can read the prologix adapter itself.

**Comment by bmoneke** [comment]
Feel free to merge, if you agree, @msmttchr

**Comment by msmttchr** [comment]
> Thanks for your review @msmttchr I changed the docstrings to pymeasure style ("Control", "Get"...), at the same time, I made the properties accept useful settings (True/False or a termination string).
> 
> As I wrote the tests, I noticed, that I have to adjust the `read` method as well: I think, if you try to read the auto setting with `++auto`, you do not have to send `++read eoi`, don't you? Now the read method can read the prologix adapter itself.

Indeed, you are right. `read` and `write` method are normally dedicated to drive instruments.
From a different perspective, it is possible to define a set of API dedicated to the adapter like:
- `write_adapter`: writes to adapter. In the current implementation, this is achieved using the regular `write` and the `write_adapter` is performed when the data to be written starts with '++'.
- 'read_adapter': reads from adapter. This is achieved with some extra parameter called `prologix`.

I would be inclined to think that this proposal has the drawback to add two more methods, but it would be more clean. In any case not a strong opinion, I just wanted to double check your opinion.

**Comment by bmoneke** [comment]
For logging purposes, read/write_adapter have to pass through the normal read/write methods and we need this `prologix` parameter anyway, so:
~~~python
def read_prologix(self):
    """Read the prologix adapter itself."""
    return self.read(prologix=True)

def write_prologix(self, command):
    """Write a command to the prologix adapter itself, prepending '++'."""
    self.write("++" + command)
~~~

If these additional methods suits you, I'll add the two methods and make all the internal commands (auto, eoi, "++ read"...) use these new methods.

I'm indifferent for their addition.

**Comment by msmttchr** [comment]
> 

Thanks for your kind feedback, I am voting for leaving things as they are, so I will merge it.

**Review by msmttchr — APPROVED** [review]
Ok for me. Thanks

**Pull request merged**